src/librustc_trans/trans/mir/rvalue.rs

```diff
@@ -0,0 +1,235 @@
+                let is_float = lhs.ty.is_fp();
+                let is_signed = lhs.ty.is_signed();
+                let binop_debug_loc = DebugLoc::None;
+                let llval = match op {
+                    mir::BinOp::Add => if is_float {
+                        build::FAdd(bcx, lhs.llval, rhs.llval, binop_debug_loc)
+                    } else {
+                        build::Add(bcx, lhs.llval, rhs.llval, binop_debug_loc)
+                    },
+                    mir::BinOp::Sub => if is_float {
+                        build::FSub(bcx, lhs.llval, rhs.llval, binop_debug_loc)
+                    } else {
+                        build::Sub(bcx, lhs.llval, rhs.llval, binop_debug_loc)
+                    },
+                    mir::BinOp::Mul => if is_float {
+                        build::FMul(bcx, lhs.llval, rhs.llval, binop_debug_loc)
+                    } else {
+                        build::Mul(bcx, lhs.llval, rhs.llval, binop_debug_loc)
+                    },
+                    mir::BinOp::Div => if is_float {
+                        build::FDiv(bcx, lhs.llval, rhs.llval, binop_debug_loc)
+                    } else if is_signed {
+                        build::SDiv(bcx, lhs.llval, rhs.llval, binop_debug_loc)
+                    } else {
+                        build::UDiv(bcx, lhs.llval, rhs.llval, binop_debug_loc)
+                    },
+                    mir::BinOp::Rem => if is_float {
+                        // LLVM currently always lowers the `frem` instructions appropriate
+                        // library calls typically found in libm. Notably f64 gets wired up
+                        // to `fmod` and f32 gets wired up to `fmodf`. Inconveniently for
+                        // us, 32-bit MSVC does not actually have a `fmodf` symbol, it's
+                        // instead just an inline function in a header that goes up to a
+                        // f64, uses `fmod`, and then comes back down to a f32.
+                        //
+                        // Although LLVM knows that `fmodf` doesn't exist on MSVC, it will
+                        // still unconditionally lower frem instructions over 32-bit floats
+                        // to a call to `fmodf`. To work around this we special case MSVC
+                        // 32-bit float rem instructions and instead do the call out to
+                        // `fmod` ourselves.
+                        //
+                        // Note that this is currently duplicated with src/libcore/ops.rs
+                        // which does the same thing, and it would be nice to perhaps unify
+                        // these two implementations on day! Also note that we call `fmod`
+                        // for both 32 and 64-bit floats because if we emit any FRem
+                        // instruction at all then LLVM is capable of optimizing it into a
+                        // 32-bit FRem (which we're trying to avoid).
```

Comment:
Should this special case exist here? I can’t find a similar case in hir trans code. Perhaps would be better to leave it out to match what HIR trans does if it does not have such case? It can always be added later too.